File: Sources/ScreenPlus/InputSounds.swift
import AVFoundation

/// Synthesized mouse-click and keyboard sounds, and the audio tracks built from them.
///
/// Everything is generated in code (no audio assets). Randomness is seeded so a recording
/// always sounds the same on every preview and export.
enum InputSounds {
    static let sampleRate = 48_000.0

    struct Hit {
        var time: Double
        var sound: [Float]
    }

    // MARK: Events → hits

    static func clickHits(_ session: RecordingSession) -> [Hit] {
        session.clicks.map { Hit(time: $0.t, sound: mouseClick) }
    }

    static func keyHits(_ session: RecordingSession) -> [Hit] {
        (session.keys ?? []).enumerated().map { index, key in
            let sound: [Float]
            switch key.kind {
            case .space: sound = spaceKey
            case .enter: sound = enterKey
            case .delete: sound = deleteKey
            case .regular: sound = regularKeys[(index * 7 + 3) % regularKeys.count]
            }
            return Hit(time: key.t, sound: sound)
        }
    }

    // MARK: Sounds

    /// Sharp press "snap" followed by a softer release, like a real mouse button.
    static let mouseClick: [Float] = {
        var out = buffer(seconds: 0.13)
        var rng = Noise(seed: 0xC11C)
        mix(into: &out, at: 0, gain: 1.0) { t in
            let snap = rng.highpassed() * exp(-t * 2_600) * 1.4
            let ping = sin(2 * .pi * 5_200 * t) * exp(-t * 1_900) * 0.8
            let tick = sin(2 * .pi * 2_700 * t) * exp(-t * 1_000) * 0.6
            let body = sin(2 * .pi * 950 * t) * exp(-t * 380) * 0.35
            return snap + ping + tick + body
        }
        mix(into: &out, at: 0.075, gain: 0.5) { t in
            let snap = rng.highpassed() * exp(-t * 3_200) * 1.2
            let ping = sin(2 * .pi * 4_600 * t) * exp(-t * 2_200) * 0.7
            let body = sin(2 * .pi * 1_100 * t) * exp(-t * 500) * 0.3
            return snap + ping + body
        }
        return normalized(out, peak: 0.95)
    }()

    /// A handful of slightly different mechanical key sounds so typing doesn't sound robotic.
    static let regularKeys: [[Float]] = (0..<6).map { i in
        let v = Double(i)
        return keySound(seed: UInt32(100 + i), thock: 210 + v * 14, plate: 1_750 + v * 90,
                        release: 0.085 + v * 0.004, peak: 0.55 + Float(i % 3) * 0.05)
    }
    static let spaceKey = keySound(seed: 7, thock: 140, plate: 1_300, release: 0.11, peak: 0.7, length: 0.24)
    static let enterKey = keySound(seed: 8, thock: 170, plate: 1_500, release: 0.1, peak: 0.75, length: 0.22)
    static let deleteKey = keySound(seed: 9, thock: 195, plate: 1_650, release: 0.09, peak: 0.6)

    /// Key press: a click of the switch, the "thock" of bottoming out, and a ring from the plate;
    /// then a quieter key release.
    private static func keySound(seed: UInt32, thock: Double, plate: Double, release: Double,
                                 peak: Float, length: Double = 0.2) -> [Float] {
        var out = buffer(seconds: length)
        var rng = Noise(seed: seed)
        mix(into: &out, at: 0, gain: 1.0) { t in
            let click = rng.highpassed() * exp(-t * 1_500) * 0.9
            let bottom = sin(2 * .pi * thock * t) * exp(-t * 55) * 0.9
            let ring = sin(2 * .pi * plate * t) * exp(-t * 420) * 0.3
            return click + bottom + ring
        }
        mix(into: &out, at: release, gain: 0.35) { t in
            let click = rng.highpassed() * exp(-t * 2_000) * 0.8
            let bottom = sin(2 * .pi * thock * 1.25 * t) * exp(-t * 80) * 0.6
            return click + bottom
        }
        return normalized(out, peak: peak)
    }

    // MARK: Track writing

    /// Writes an AAC (.m4a) track of `duration` seconds with every hit mixed in at its time.
    static func writeTrack(_ hits: [Hit], duration: Double, to url: URL) throws {
        try? FileManager.default.removeItem(at: url)
        let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
        let file = try AVAudioFile(forWriting: url, settings: [
            AVFormatIDKey: kAudioFormatMPEG4AAC,
            AVSampleRateKey: sampleRate,
            AVNumberOfChannelsKey: 1,
            AVEncoderBitRateKey: 128_000,
        ], commonFormat: .pcmFormatFloat32, interleaved: false)

        let totalFrames = Int(max(duration, 0.1) * sampleRate)
        let placed = hits.filter { $0.time >= 0 && $0.time < duration }
            .map { (start: Int($0.time * sampleRate), sound: $0.sound) }
            .sorted { $0.start < $1.start }
        let chunk = Int(sampleRate)  // one second at a time

        var chunkStart = 0
        while chunkStart < totalFrames {
            let count = min(chunk, totalFrames - chunkStart)
            let pcm = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(count))!
            pcm.frameLength = AVAudioFrameCount(count)
            let data = pcm.floatChannelData![0]
            data.initialize(repeating: 0, count: count)

            for hit in placed where hit.start < chunkStart + count && hit.start + hit.sound.count > chunkStart {
                let from = max(chunkStart, hit.start), to = min(chunkStart + count, hit.start + hit.sound.count)
                for frame in from..<to {
                    data[frame - chunkStart] += hit.sound[frame - hit.start]
                }
            }
            for i in 0..<count { data[i] = max(-1, min(1, data[i])) }  // overlapping hits can't clip
            try file.write(from: pcm)
            chunkStart += count
        }
    }

    /// Combines a video file with an audio track into an .mp4 without re-encoding the video.
    static func mux(video: URL, audio: URL, to output: URL) async throws {
        let (composition, _) = try await combine(video: AVURLAsset(url: video), audio: [audio])
        try? FileManager.default.removeItem(at: output)
        guard let export = AVAssetExportSession(asset: composition, presetName: AVAssetExportPresetPassthrough) else {
            throw RenderError.failed("could not create audio export session")
        }
        try await export.export(to: output, as: .mp4)
    }

    /// A composition with the video's picture plus one audio track per file.
    /// Returns the audio track IDs in the same order, so the preview can mute them individually.
    static func combine(video videoAsset: AVAsset, audio: [URL]) async throws
        -> (AVMutableComposition, [CMPersistentTrackID])
    {
        let composition = AVMutableComposition()
        let duration = try await videoAsset.load(.duration)

        if let track = try await videoAsset.loadTracks(withMediaType: .video).first,
           let target = composition.addMutableTrack(withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid) {
            try target.insertTimeRange(CMTimeRange(start: .zero, duration: duration), of: track, at: .zero)
        }

        var audioIDs: [CMPersistentTrackID] = []
        for url in audio {
            let asset = AVURLAsset(url: url)
            guard let track = try await asset.loadTracks(withMediaType: .audio).first,
                  let target = composition.addMutableTrack(withMediaType: .audio,
                                                           preferredTrackID: kCMPersistentTrackID_Invalid)
            else { continue }
            let audioDuration = try await asset.load(.duration)
            try target.insertTimeRange(CMTimeRange(start: .zero, duration: min(duration, audioDuration)),
                                       of: track, at: .zero)
            audioIDs.append(target.trackID)
        }
        return (composition, audioIDs)
    }

    static func temporaryURL(_ name: String) -> URL {
        FileManager.default.temporaryDirectory.appendingPathComponent("ScreenPlus-\(name)-\(UUID().uuidString).m4a")
    }

    // MARK: Synthesis helpers

    private static func buffer(seconds: Double) -> [Float] {
        [Float](repeating: 0, count: Int(sampleRate * seconds))
    }

    private static func mix(into out: inout [Float], at offset: Double, gain: Double, _ voice: (Double) -> Double) {
        let start = Int(offset * sampleRate)
        for i in start..<out.count {
            out[i] += Float(voice(Double(i - start) / sampleRate) * gain)
        }
    }

    private static func normalized(_ samples: [Float], peak: Float) -> [Float] {
        let maxValue = samples.map(abs).max() ?? 1
        guard maxValue > 0 else { return samples }
        return samples.map { $0 / maxValue * peak }
    }

    /// Deterministic white noise, with a first-order high-pass for a brighter "snap".
    private struct Noise {
        var state: UInt32
        var previous: Double = 0

        init(seed: UInt32) { state = seed &* 2_654_435_761 | 1 }

        mutating func next() -> Double {
            state = state &* 1_664_525 &+ 1_013_904_223
            return Double(state >> 8) / Double(1 << 24) * 2 - 1
        }

        mutating func highpassed() -> Double {
            let x = next()
            defer { previous = x }
            return (x - previous) * 0.5
        }
    }
}

File: windows/src/ScreenPlus.Core/InputSounds.cs
namespace ScreenPlus;

/// <summary>
/// Synthesized mouse-click and keyboard sounds, and the audio tracks built from them.
///
/// Everything is generated in code (no audio assets). Randomness is seeded so a recording
/// always sounds the same on every preview and export.
/// </summary>
public static class InputSounds
{
    public const int SampleRate = 48_000;

    public readonly record struct Hit(double Time, float[] Sound);

    // MARK: Events → hits

    public static List<Hit> ClickHits(RecordingSession session) =>
        session.Clicks.Select(c => new Hit(c.T, MouseClick)).ToList();

    public static List<Hit> KeyHits(RecordingSession session) =>
        (session.Keys ?? []).Select((key, index) => new Hit(key.T, key.Kind switch
        {
            KeyKind.Space => SpaceKey,
            KeyKind.Enter => EnterKey,
            KeyKind.Delete => DeleteKey,
            _ => RegularKeys[(index * 7 + 3) % RegularKeys.Length],
        })).ToList();

    /// <summary>
    /// Moves hits onto the output timeline of a sped-up or slowed-down video. The sounds themselves keep
    /// their length and pitch: a click still sounds like a click.
    /// </summary>
    public static List<Hit> AtSpeed(IEnumerable<Hit> hits, double speed) =>
        hits.Select(h => h with { Time = h.Time / speed }).ToList();

    // MARK: Sounds

    /// <summary>Sharp press "snap" followed by a softer release, like a real mouse button.</summary>
    public static readonly float[] MouseClick = MakeMouseClick();

    /// <summary>A handful of slightly different mechanical key sounds so typing doesn't sound robotic.</summary>
    public static readonly float[][] RegularKeys = Enumerable.Range(0, 6).Select(i =>
    {
        double v = i;
        return KeySound(seed: (uint)(100 + i), thock: 210 + v * 14, plate: 1_750 + v * 90,
                        release: 0.085 + v * 0.004, peak: 0.55f + (i % 3) * 0.05f);
    }).ToArray();

    public static readonly float[] SpaceKey = KeySound(seed: 7, thock: 140, plate: 1_300, release: 0.11, peak: 0.7f, length: 0.24);
    public static readonly float[] EnterKey = KeySound(seed: 8, thock: 170, plate: 1_500, release: 0.1, peak: 0.75f, length: 0.22);
    public static readonly float[] DeleteKey = KeySound(seed: 9, thock: 195, plate: 1_650, release: 0.09, peak: 0.6f);

    private static float[] MakeMouseClick()
    {
        var output = Buffer(seconds: 0.13);
        var rng = new Noise(0xC11C);
        Mix(output, 0, 1.0, t =>
        {
            var snap = rng.HighPassed() * Math.Exp(-t * 2_600) * 1.4;
            var ping = Math.Sin(2 * Math.PI * 5_200 * t) * Math.Exp(-t * 1_900) * 0.8;
            var tick = Math.Sin(2 * Math.PI * 2_700 * t) * Math.Exp(-t * 1_000) * 0.6;
            var body = Math.Sin(2 * Math.PI * 950 * t) * Math.Exp(-t * 380) * 0.35;
            return snap + ping + tick + body;
        });
        Mix(output, 0.075, 0.5, t =>
        {
            var snap = rng.HighPassed() * Math.Exp(-t * 3_200) * 1.2;
            var ping = Math.Sin(2 * Math.PI * 4_600 * t) * Math.Exp(-t * 2_200) * 0.7;
            var body = Math.Sin(2 * Math.PI * 1_100 * t) * Math.Exp(-t * 500) * 0.3;
            return snap + ping + body;
        });
        return Normalized(output, peak: 0.95f);
    }

    /// <summary>
    /// Key press: a click of the switch, the "thock" of bottoming out, and a ring from the plate;
    /// then a quieter key release.
    /// </summary>
    private static float[] KeySound(uint seed, double thock, double plate, double release, float peak, double length = 0.2)
    {
        var output = Buffer(seconds: length);
        var rng = new Noise(seed);
        Mix(output, 0, 1.0, t =>
        {
            var click = rng.HighPassed() * Math.Exp(-t * 1_500) * 0.9;
            var bottom = Math.Sin(2 * Math.PI * thock * t) * Math.Exp(-t * 55) * 0.9;
            var ring = Math.Sin(2 * Math.PI * plate * t) * Math.Exp(-t * 420) * 0.3;
            return click + bottom + ring;
        });
        Mix(output, release, 0.35, t =>
        {
            var click = rng.HighPassed() * Math.Exp(-t * 2_000) * 0.8;
            var bottom = Math.Sin(2 * Math.PI * thock * 1.25 * t) * Math.Exp(-t * 80) * 0.6;
            return click + bottom;
        });
        return Normalized(output, peak);
    }

    // MARK: Synthesis helpers

    private static float[] Buffer(double seconds) => new float[(int)(SampleRate * seconds)];

    private static void Mix(float[] output, double offset, double gain, Func<double, double> voice)
    {
        var start = (int)(offset * SampleRate);
        for (var i = start; i < output.Length; i++)
            output[i] += (float)(voice((double)(i - start) / SampleRate) * gain);
    }

    private static float[] Normalized(float[] samples, float peak)
    {
        var maxValue = samples.Length == 0 ? 1 : samples.Max(MathF.Abs);
        if (maxValue <= 0) return samples;
        return samples.Select(s => s / maxValue * peak).ToArray();
    }

    /// <summary>Deterministic white noise, with a first-order high-pass for a brighter "snap".</summary>
    private struct Noise(uint seed)
    {
        private uint _state = unchecked(seed * 2_654_435_761u) | 1;
        private double _previous = 0;

        public double Next()
        {
            _state = unchecked(_state * 1_664_525u + 1_013_904_223u);
            return (double)(_state >> 8) / (1 << 24) * 2 - 1;
        }

        public double HighPassed()
        {
            var x = Next();
            var result = (x - _previous) * 0.5;
            _previous = x;
            return result;
        }
    }
}

/// <summary>
/// A mono track of sound hits placed on a timeline, mixed on demand. Used both for streaming
/// preview audio and for the exported audio track.
/// </summary>
public sealed class SoundTrack
{
    private readonly (long Start, float[] Sound)[] _hits;
    private readonly int _longest;

    public SoundTrack(IEnumerable<InputSounds.Hit> hits, double duration)
    {
        _hits = hits.Where(h => h.Time >= 0 && h.Time < duration)
            .Select(h => (Start: (long)(h.Time * InputSounds.SampleRate), h.Sound))
            .OrderBy(h => h.Start)
            .ToArray();
        _longest = _hits.Length == 0 ? 0 : _hits.Max(h => h.Sound.Length);
    }

    public bool IsEmpty => _hits.Length == 0;

    /// <summary>Adds the track's samples for <c>[startSample, startSample + destination.Length)</c> into <paramref name="destination"/>.</summary>
    public void MixInto(long startSample, Span<float> destination)
    {
        if (_hits.Length == 0 || destination.IsEmpty) return;
        var end = startSample + destination.Length;
        // First hit that can still be sounding at startSample.
        long from = startSample - _longest;
        int lo = 0, hi = _hits.Length;
        while (lo < hi)
        {
            var mid = (lo + hi) / 2;
            if (_hits[mid].Start < from) lo = mid + 1; else hi = mid;
        }
        for (var i = lo; i < _hits.Length && _hits[i].Start < end; i++)
        {
            var (start, sound) = _hits[i];
            var a = Math.Max(startSample, start);
            var b = Math.Min(end, start + sound.Length);
            for (var s = a; s < b; s++)
                destination[(int)(s - startSample)] += sound[s - start];
        }
    }

    /// <summary>Clamps mixed samples so overlapping hits can't clip, and converts them to 16-bit PCM.</summary>
    public static void ToPcm16(ReadOnlySpan<float> samples, Span<short> destination)
    {
        for (var i = 0; i < samples.Length; i++)
            destination[i] = (short)Math.Round(Math.Clamp(samples[i], -1f, 1f) * short.MaxValue);
    }
}
